### DISCUSSION TITLE
Pagination losses url

### DISCUSSION THREAD

**Discussion by ChrisSantiago82** [livewire/livewire#3079] [7. Bugs (archived)]
I'm using the latest version of Livewire.

If I want use pagination with Livewire\WithPagination and  use WithPagination, and if I visit that site, on the top the url gets replaces with "livewire/message/......". And then when the pagination should do something of course I'm getting a MethodNotAllowedHttpException.

url with pagination enabled:
![image](https://user-images.githubusercontent.com/43000085/96003554-55fb6c80-0e08-11eb-9179-c6891f6e419c.png)

url without pagination (same page):
![image](https://user-images.githubusercontent.com/43000085/96003627-6a3f6980-0e08-11eb-9a57-742cb112759e.png)

Code:
`<?php

namespace App\Http\Livewire\Order;

use Livewire\Component;
use Livewire\WithPagination;
use App\Bearer;

class Insurancemodal extends Component
{

    use WithPagination;

    protected $paginationTheme = 'bootstrap';

    public $search;

    public function updatingSearch()
    {
        $this->resetPage();
    }

    public function render()
    {

        return view('livewire.order.insurancemodal', [
            'bearers' => Bearer::where('company', 'like', '%'.$this->search.'%')->paginate(10),
        ]);

    }
}`

**Comment by ChrisThompsonTLDR** [comment]
By chance, are you using `wire:click.prevent.prefetch` on pagination/ordering/filtering links?  Or `url()->current()`?

We can replicate this issue with the following:
```
<a href="{{ url()->current() . '?' . http_build_query(['limit' => $limit, 'orderBy' => $name, 'orderDir' => (($orderDir === 'desc') ? 'asc' : 'desc')]) }}" wire:click.prevent.prefetch="order('{{ $name }}', '{{ $orderDir === 'desc' ? 'asc' : 'desc' }}')">
```

If we change to the following we don't experience the issue:
```
<a href="{{ url()->current() . '?' . http_build_query(['limit' => $limit, 'orderBy' => $name, 'orderDir' => (($orderDir === 'desc') ? 'asc' : 'desc')]) }}" wire:click="order('{{ $name }}', '{{ $orderDir === 'desc' ? 'asc' : 'desc' }}')">
```

**Comment by ChrisThompsonTLDR** [comment]
Are you on Laravel 7 and Livewire 2?

**Comment by ChrisSantiago82** [comment]
I don't think so.
It happens when the component the first time is mounted that the browser is change the url.
I have tried it out now without any query that would load a pagination. Just the "use WithPagination" will change the url.
Yes, I'm on Laravel 7 with Livewire 2.2.9

**Comment by squishythejellyfish** [comment]
👋 Oh Hi! I'm Squishy, the friendly jellyfish that manages Livewire issues.

I see this issue has been closed.

Here in the Livewire repo, we have an "issues can be closed guilt-free and without explanation" policy.

If for ANY reason you think this issue hasn't been resolved, PLEASE feel empowered to re-open it.

Re-opening actually helps us track which issues are a priority.

Reply "REOPEN" to this comment and we'll happily re-open it for you!

(More info on this philosophy here: https://twitter.com/calebporzio/status/1321864801295978497)

**Comment by ChrisSantiago82** [comment]
Hey, 
I can't understand this philosophy and I'm feeling that I'm have been silenced. 
Don't get me wrong, I think it's a great project that has grown a lot in a short time. 

I also think I'm a person that can understand a lot.
I could understand if you say you need more information.  I could understand if you say that there is no capacity to resolve it. I could understand if you say that there are not enough people affected by this problem. I could understand if you would ask for money to resolve this issue. I could understand if it's on the lowest priority. 
But just closing it without a comment doesn't give a good feeling. Maybe I also didn't described it well enough and I could accept such an answer as well. 
But with the next issue I could think why bother creating an issue. And maybe other people will think this as well. But I also think the quality of the project has improved alot just of all the issues that have been reported. If people feel like it's not worth it creating issues.. It's a slippery slope.

**Comment by calebporzio** [comment]
I declared issue bankruptcy and closed every single issue without comment.

I created a bot to allow anyone at all to re-open an issue, specifically so that people WOULDN'T feel silenced.

Please re-open this issue if you want more attention brought to it.

**Comment by ChrisSantiago82** [comment]
I have the same issue again with another project. This time it's with Laravel 8. I couldn't find a way to reopen this issue. Do you want me to open a new one?

**Comment by iksaku** [comment]
REOPEN

**Comment by iksaku** [comment]
This is indeed still an issue, but thanks to new Laravel documentation, I Think I got a way to get around this.

Every paginator instance can take a [custom path](https://laravel.com/docs/8.x/pagination#customizing-pagination-urls 
) and assign it to the provided URLs of the `links()` method.

I think at the end you could could opt-in for this feature in full-page components, or manually assign it for partial components.

Could be part of the existing WithPagination trait, behind a `protected $trackComponentPath` property.

What do you guys think?

**Comment by austenc** [comment]
If someone provides a demo repository that allows me to easily reproduce this bug, I'd be happy to take a look at it.

@iksaku perhaps you have one?

Thanks!

**Comment by ChrisSantiago82** [comment]
@austenc 

Here is a demo repository:
https://github.com/ChrisSantiago82/Livewire-Test

When you click on the Pagination link in the sidebar, you can see that the url changes in the browser. And of course then the pagination links go to the wrong url.
If you remove the pagination from the component, then the url doesn't change when you open the link.

**Comment by austenc** [comment]
Thank you @ChrisSantiago82 -- I see the issue, but I'm not sure it's an easy one to fix. Since you're loading the component dynamically, any modification to the query string (which `WithPagination` does) will cause this behavior. 

The tricky part is that in this case, sometimes we want livewire to send path updates (like when actually clicking a "Next" pagination link), and sometimes we don't (like when initializing a component dynamically).

**Comment by ChrisSantiago82** [comment]
@austenc Thanks a lot investigating it. I can understand the problem. Lately I'm using livewire more and more and I also have created applications almost like a single page application. The sidebar has dropdowns and selected items, and if I just create that as a livewire component and load the other components dynamically, I don't have to save the state of the side bar.

I can imagine that other people also load components dynamically and I assume they would have the same problems.
If I'm the only one that has the problem, I would just program my own pagination but if other people have the same issues then we should do something. 
You have written that the problem is that livewire "wouldn't know" if the component was loaded dynamically or not. Maybe the user could set an additional parameter manually when it's loaded dynamically or we could create a "WithPagination" for dynamic loaded component that would have a different name. Or as iksaku has mentioned, maybe we could set the links parameter or something like this. What do you think?

**Comment by rmrfstern** [comment]
I have the same issue. Loading dynamic a component, which includes Pagination, the url changes to .../livewire/message/... . I am new to laravel, so I have no idea how to work around.

**Comment by joshhanley** [comment]
I think #2333 is related

**Comment by alisanie** [comment]
I have a same problem but just for now I suggest to add `use WithPagination` in you **parent component** too and It's working.
The Problem is I don't think its a good way

**Comment by OscarKolsrud** [comment]
> I have a same problem but just for now I suggest to add `use WithPagination` in you **parent component** too and It's working.
> The Problem is I don't think its a good way

You my dear man are a god, the clock is now 01:30 and this is the solution as of now.

@calebporzio Would it be an idea to add this workaround to the documentation until a "proper" fix is implemented?

**Comment by aliusmn** [comment]
> I have a same problem but just for now I suggest to add `use WithPagination` in you **parent component** too and It's working.
> The Problem is I don't think its a good way

Thanks, you saved me

**Comment by ChrisSantiago82** [comment]
Yes, that seams to do the trick. Thank you very much @alisanie 
Now I can update from version 1 to version 2.
We even could consider closing this issue if we write it into the documentation.

**Comment by FaycalBorsali** [comment]
> I have a same problem but just for now I suggest to add `use WithPagination` in you **parent component** too and It's working.
> The Problem is I don't think its a good way

THANK YOU, it indeed saved my day !

**Comment by Jam0r85** [comment]
Having the same issue without a parent livewire component.

The controller returns a simple view.
The view contains a single livewire component

Contents of the livewire component are as follows

```<?php

namespace App\Http\Livewire\Admin\Property;

use App\Models\Property;
use Livewire\Component;
use Livewire\WithPagination;

class SearchEicrProperties extends Component
{
    use WithPagination;

    public $search;

    public $status;

    public $managed;

    protected $listeners = [
        'eicrRecorded' => '$refresh'
    ];

    public function mount()
    {
        $this->search = request()->query('search', $this->search);
        $this->status = request()->query('status', $this->status);
        $this->managed = request()->query('managed', $this->managed);
    }

    public function recordEicr($propertyId)
    {
        $this->emit('recordEicr', $propertyId);
    }

    public function updated()
    {
        $this->reset('page');
    }

    public function render()
    {
        return view('livewire.admin.property.search-eicr-properties', [
            'properties' => Property::filterWithLetting([
                'search' => $this->search,
                'status' => $this->status,
                'managed' => $this->managed
            ])->latest()->paginate()
        ]);
    }
}
```

Whenever I update any of the public properties and try to use the pagination links it contains the error.

**Comment by Jam0r85** [comment]
I've just removed my tailwind pagination view and it's resolved the issue?

```
@if ($paginator->hasPages())
    <nav role="navigation" aria-label="{{ __('Pagination Navigation') }}" class="flex items-center justify-between mt-2">
        <div class="flex justify-between flex-1 sm:hidden">
            @if ($paginator->onFirstPage())
                <span class="relative inline-flex items-center px-4 py-2 text-sm font-medium text-gray-500 bg-white border border-gray-300 cursor-default leading-5 rounded-md">
                    {!! __('pagination.previous') !!}
                </span>
            @else
                <a href="{{ $paginator->previousPageUrl() }}" class="relative inline-flex items-center px-4 py-2 text-sm font-medium text-gray-700 bg-white border border-gray-300 leading-5 rounded-md hover:text-gray-500 focus:outline-none focus:shadow-outline-blue focus:border-blue-300 active:bg-gray-100 active:text-gray-700 transition ease-in-out duration-150">
                    {!! __('pagination.previous') !!}
                </a>
            @endif

            @if ($paginator->hasMorePages())
                <a href="{{ $paginator->nextPageUrl() }}" class="relative inline-flex items-center px-4 py-2 ml-3 text-sm font-medium text-gray-700 bg-white border border-gray-300 leading-5 rounded-md hover:text-gray-500 focus:outline-none focus:shadow-outline-blue focus:border-blue-300 active:bg-gray-100 active:text-gray-700 transition ease-in-out duration-150">
                    {!! __('pagination.next') !!}
                </a>
            @else
                <span class="relative inline-flex items-center px-4 py-2 ml-3 text-sm font-medium text-gray-500 bg-white border border-gray-300 cursor-default leading-5 rounded-md">
                    {!! __('pagination.next') !!}
                </span>
            @endif
        </div>

        <div class="hidden sm:flex-1 sm:flex sm:items-center sm:justify-between">
            <div>
                <p class="text-sm text-gray-400 leading-5">
                    {!! __('Showing') !!}
                    <span class="font-medium">{{ $paginator->firstItem() }}</span>
                    {!! __('to') !!}
                    <span class="font-medium">{{ $paginator->lastItem() }}</span>
                    {!! __('of') !!}
                    <span class="font-medium">{{ $paginator->total() }}</span>
                    {!! __('results') !!}
                </p>
            </div>

            <div>
                <span class="relative z-0 inline-flex shadow-sm rounded-md">
                    {{-- Previous Page Link --}}
                    @if ($paginator->onFirstPage())
                        <span aria-disabled="true" aria-label="{{ __('pagination.previous') }}">
                            <span class="relative inline-flex items-center px-2 py-2 text-sm font-medium text-gray-500 bg-white border border-gray-300 cursor-default rounded-l-md leading-5" aria-hidden="true">
                                <svg class="w-5 h-5" fill="currentColor" viewBox="0 0 20 20">
                                    <path fill-rule="evenodd" d="M12.707 5.293a1 1 0 010 1.414L9.414 10l3.293 3.293a1 1 0 01-1.414 1.414l-4-4a1 1 0 010-1.414l4-4a1 1 0 011.414 0z" clip-rule="evenodd" />
                                </svg>
                            </span>
                        </span>
                    @else
                        <a href="{{ $paginator->previousPageUrl() }}" rel="prev" class="relative inline-flex items-center px-2 py-2 text-sm font-medium text-gray-500 bg-white border border-gray-300 rounded-l-md leading-5 hover:text-gray-400 focus:z-10 focus:outline-none focus:border-blue-300 focus:shadow-outline-blue active:bg-gray-100 active:text-gray-500 transition ease-in-out duration-150" aria-label="{{ __('pagination.previous') }}">
                            <svg class="w-5 h-5" fill="currentColor" viewBox="0 0 20 20">
                                <path fill-rule="evenodd" d="M12.707 5.293a1 1 0 010 1.414L9.414 10l3.293 3.293a1 1 0 01-1.414 1.414l-4-4a1 1 0 010-1.414l4-4a1 1 0 011.414 0z" clip-rule="evenodd" />
                            </svg>
                        </a>
                    @endif

                    {{-- Pagination Elements --}}
                    @foreach ($elements as $element)
                        {{-- "Three Dots" Separator --}}
                        @if (is_string($element))
                            <span aria-disabled="true">
                                <span class="relative inline-flex items-center px-4 py-2 -ml-px text-sm font-medium text-gray-700 bg-white border border-gray-300 cursor-default leading-5">{{ $element }}</span>
                            </span>
                        @endif

                        {{-- Array Of Links --}}
                        @if (is_array($element))
                            @foreach ($element as $page => $url)
                                @if ($page == $paginator->currentPage())
                                    <span aria-current="page">
                                        <span class="relative inline-flex items-center px-4 py-2 -ml-px text-sm font-medium text-gray-500 bg-white border border-gray-300 cursor-default leading-5">{{ $page }}</span>
                                    </span>
                                @else
                                    <a href="{{ $url }}" class="relative inline-flex items-center px-4 py-2 -ml-px text-sm font-medium text-gray-700 bg-white border border-gray-300 leading-5 hover:text-gray-500 focus:z-10 focus:outline-none focus:border-blue-300 focus:shadow-outline-blue active:bg-gray-100 active:text-gray-700 transition ease-in-out duration-150" aria-label="{{ __('Go to page :page', ['page' => $page]) }}">
                                        {{ $page }}
                                    </a>
                                @endif
                            @endforeach
                        @endif
                    @endforeach

                    {{-- Next Page Link --}}
                    @if ($paginator->hasMorePages())
                        <a href="{{ $paginator->nextPageUrl() }}" rel="next" class="relative inline-flex items-center px-2 py-2 -ml-px text-sm font-medium text-gray-500 bg-white border border-gray-300 rounded-r-md leading-5 hover:text-gray-400 focus:z-10 focus:outline-none focus:border-blue-300 focus:shadow-outline-blue active:bg-gray-100 active:text-gray-500 transition ease-in-out duration-150" aria-label="{{ __('pagination.next') }}">
                            <svg class="w-5 h-5" fill="currentColor" viewBox="0 0 20 20">
                                <path fill-rule="evenodd" d="M7.293 14.707a1 1 0 010-1.414L10.586 10 7.293 6.707a1 1 0 011.414-1.414l4 4a1 1 0 010 1.414l-4 4a1 1 0 01-1.414 0z" clip-rule="evenodd" />
                            </svg>
                        </a>
                    @else
                        <span aria-disabled="true" aria-label="{{ __('pagination.next') }}">
                            <span class="relative inline-flex items-center px-2 py-2 -ml-px text-sm font-medium text-gray-500 bg-white border border-gray-300 cursor-default rounded-r-md leading-5" aria-hidden="true">
                                <svg class="w-5 h-5" fill="currentColor" viewBox="0 0 20 20">
                                    <path fill-rule="evenodd" d="M7.293 14.707a1 1 0 010-1.414L10.586 10 7.293 6.707a1 1 0 011.414-1.414l4 4a1 1 0 010 1.414l-4 4a1 1 0 01-1.414 0z" clip-rule="evenodd" />
                                </svg>
                            </span>
                        </span>
                    @endif
                </span>
            </div>
        </div>
    </nav>
@endif

```

**Comment by SanderMuller** [comment]
Also having this issue without a parent livewire component just like @Jam0r85.
Has anyone found a solution for this issue?

**Comment by chimerakraken** [comment]
> Also having this issue without a parent livewire component just like @Jam0r85.
> Has anyone found a solution for this issue?

you can try to use `@include` to the child component rather than` @livewire` to prevent the child pagination from affecting mother component

**Reply by SanderMuller** [reply]
> > Also having this issue without a parent livewire component just like @Jam0r85.
> > Has anyone found a solution for this issue?
> 
> you can try to use `@include` to the child component rather than` @livewire` to prevent the child pagination from affecting mother component

I am not using a child component, just a single component with pagination

**Comment by Liam-PVTL** [comment]
> > Also having this issue without a parent livewire component just like @Jam0r85.
> > Has anyone found a solution for this issue?
> 
> you can try to use `@include` to the child component rather than` @livewire` to prevent the child pagination from affecting mother component

This is hard when using directly loaded components from a route, e.g

`Route::get('roles', RolesBrowse::class)->name('roles.browse');`

**Comment by Bishal0009** [comment]
Try replacing href by `wire:click.prevent`, It solves for me. For example,
` 
<button wire:click.prevent="previousPage" wire:loading.attr="disabled" rel="prev" class="btn">
{!! __('pagination.previous') !!}
</button>
`
Similarly, use `wire:click.prevent="gotoPage({{ $page }})"` and  `wire:click.prevent="nextPage"`

**Comment by NehalPatel** [comment]
Default Pagination doesn't have this issue anymore, but still this issue with bootstrap-5 pagination view.

`{{ $records->links('pagination::bootstrap-5') }}`

After the component re-render, the pagination URLs change to 'livewire/message/{component-name}?page=1'

**Reply by joshhanley** [reply]
@NehalPatel make sure you are using Livewire's version of the bootstrap pagination, as Laravel's one (which it looks like you are trying to use) won't work.

See here in the docs for how to use it https://laravel-livewire.com/docs/2.x/pagination#custom-pagination-view

Hope this helps!

**Comment by Marcy3** [comment]
I have the same problem when I have a second component that is rendered. I have a select that displays a table, the values in the table change depending on the select value. Since the table has many entries, I try to split the values into several pages with `->paginate(50)`. But when I click on the second page, the URL changes to `http://localhost:8000/?page=2` instead of `http://localhost:8000/items/?page=2`. Without the select, the URL works as intended. 
I use the following things in the components 
`use WithPagination`
`protected $paginationTheme = 'bootstrap';`

in the views:
`<select class="form-select text-center" wire:change="selectedName($event.target.value)">`
`<div class="row" wire:loading.delay.remove wire:target="selectedName">`

I've had it both in one component and split into multiple components and still the URL changes, even when I add the trait to the parent component
Would anyone have an idea what the problem is

**Reply by Marcy3** [reply]
Problem was the wrong in the meta tags the referrer was on origin

**Comment by dawwwid** [comment]
I think I had a similar problem and I wanted to keep the links for SEO (the Google Search Console captures errors "405 Method Not Allowed"...)

To get around the problem and have seo-friendly links, I added this method to the Livewire controller:

```
/**
 * When "Paginators" enters into action, we override the function that solves the path of pagination urls
 */
public function updatingPaginators()
{
    $queries = $this->except([ 'page', 'paginators' ]);

    foreach($queries as $property => $value)
    {
        if(isset($this->queryString[ $property ]['except']) && $this->queryString[ $property ]['except'] === $value)
        {
            unset($queries[ $property ]);
        }
    }

    Paginator::currentPathResolver(function() use ($queries){
        return route('routeName', $queries);
    });
}
```
We will have to add the call to `updatingPaginators()` in `mount()` for initialization:
```
public function mount()
{
    $this->updatingPaginators();
    /* ... */
}
```
I hope it can help!

**Comment by pedestrianlove** [comment]
I was also faced with broken links that points to nonexistent route, hence cause the error that forbid the access.
Although I'm not sure if it's the right place to share my workaround on this, I'll put it here.

---

I've taken the following steps:
1. Put `use WithPagination` in both the Controller and Livewire Component Controller.
```
...
Controller
|-- View
|     |-- Livewire Component Controller
|     |      |-- Component view
...
```
That is, you don't need a parent component.

2. Copy the pagination file you desired under the livewire vendor folder to the views folder.(in my case, is `tailwind-pagination.blade.php`)
![image](https://user-images.githubusercontent.com/37164678/221114460-2622a771-6689-405d-ac56-f8d06e9ea3e9.png)
![image](https://user-images.githubusercontent.com/37164678/221114549-2568ae06-0d78-40fe-8019-18c98458cae5.png)

3. Reference it inside of `app/Providers/AppServiceProvider.php`:
```php
<?php

namespace App\Providers;

use Illuminate\Pagination\Paginator;
use Illuminate\Support\Facades\Vite;
use Illuminate\Support\ServiceProvider;

class AppServiceProvider extends ServiceProvider
{
    /**
     * Register any application services.
     *
     * @return void
     */
    public function register()
    {
        //
    }

    /**
     * Bootstrap any application services.
     *
     * @return void
     */
    public function boot()
    {
        Paginator::defaultView('components.tailwind-pagination');
        Vite::macro('image', fn (string $asset) => Vite::asset("resources/images/{$asset}"));
    }
}
```
Then it just works.

---

Hope it helps.
